crates/storage/src/lazy/mapping.rs

```diff
@@ -85,14 +85,35 @@ where
     where
         Q: scale::EncodeLike<K>,
     {
-        pull_packed_root_opt(&self.storage_key(key))
+        pull_packed_root_opt(&self.storage_key(&key))
+    }
+
+    /// Clears the value at `key` from storage.
+    ///
+    /// Returns the value at the key if it previously existed, otherwise returns `None`.
+    pub fn remove<Q>(&self, key: Q) -> Option<V>
+    where
+        Q: scale::EncodeLike<K>,
+    {
+        let storage_key = self.storage_key(&key);
+        let value = self.get(key)?;
```

Comment:
The whole point of `REQUIRES_DEEP_CLEAN_UP` is to be able to move this particular line of code inside the if-then clause with the premise that for most common types it evaluates to `false`.

Reply:
Sorry, but this doesn't make sense to me given the API I want to achieve with `remove`. In the non `REQUIRES_DEEP_CLEAN_UP` case how would we return the value back to the user?

Reply:
I would not return any value in the first place. ;)
The `Mapping` is a fundamental low-level storage facility. Therefore we should only provide APIs with which we can efficiently operate.
The `remove` method therefore should not return a value since then we'd be required to always costly load a value even though we could avoid doing so in the most common cases.
If we later find out that getting a value and then removing it (simulating a `remove` that returns the removed value) is more costly than a built-in `Mapping::remove_get` that is implemented as you currently do it, then we might add that method later on.

Reply:
Note that for example [`LazyIndexMap::clear_packed_at`](https://github.com/paritytech/ink/blob/master/crates/storage/src/lazy/lazy_imap.rs#L293) also does _not_ return the removed value for the same reason. We might even want to name both methods the same although I am not sure if `clear_packed_at` is actually a decent name.

Reply:
Thanks for clarifying how you think the API of `remove` should be. I was modelling it off `HashMap/BTreeMap::remove()` which return the removed element if it exists. This would help make `Mapping` a drop-in replacement in some of our example contracts